### DISCUSSION TITLE
Why correlation test of microeco showed different results from the same test using package microbiome?

### DISCUSSION THREAD

**Discussion by MPNguyen2022** [ChiLiubio/microeco#194] [Q&A] [answered]
Hi,

I would like to ask how the command trans_env$cal_cor exactly does for the correlation test between the abundance data and environmental data.

I tried both commands:
**cal_cor(dataset4,cor_method = "spearman") from microeco -> showed significant correlation for 3 genera**

and 
**microbiome::associate(spe, env, method = "spearman") -> showed no significant results**

I would like to ask why the microeco package gave different results and what the command actually does.

Thank you very much.

**Comment by ChiLiubio** [comment]
Hi.
The main reason is the default adjustment method in cal_cor is not same with microbiome::associate. I take an example to show how it works.
```
library(microeco)
# env alpha correlation
t1 <- trans_env$new(dataset = dataset, add_data = env_data_16S[, 4:11])
t1$cal_cor(add_abund_table = dataset$alpha_diversity, cor_method = "spearman")
# microbiome
m1 <- microbiome::associate(dataset$alpha_diversity, t1$data_env, method = "spearman")
```
The padj in m1 adjust all the raw p values. But the default method in t1 adjust each taxon individually.
```
t1$res_cor$newpadj <- p.adjust(t1$res_cor$Pvalue, "fdr")
p.adjust(t1$res_cor[t1$res_cor$Env == "Temperature", "Pvalue"], "fdr")
```
Then you can find newpadj in ` t1$res_cor` is same with that in m1. The `AdjPvalue` in t1$res_cor comes from the operation of the second way.
The key reason to design it in cal_cor is that generally there are two many taxa and environmental factors to do paired correlations. To adjust all p values together can make too many values non-significant. In my opinion, there is no standard p value adjustment way to make itself acceptable to everyone. The way with high tolerance can get more significant results that merit futher study. I want to say they are not conflicting. Please select a way you believe.
Thank you for letting me think this question again. I will add more ways in the `p_adjust_type` parameter. Maybe in the future, the default way will be same with that in microbiome package.

Best,
Chi

**Reply by ChiLiubio** [reply]
I forget to tell you that parameter `p_adjust_type = 'Type'` can make the result same with m1. Let's take an example.
```
t2 <- trans_env$new(dataset = dataset, add_data = env_data_16S[, 4:11])
t2$cal_cor(add_abund_table = dataset$alpha_diversity, cor_method = "spearman", p_adjust_type = "Type")
t2$res_cor$newpadj <- p.adjust(t2$res_cor$Pvalue, "fdr")
View(t2$res_cor)
m2 <- microbiome::associate(dataset$alpha_diversity, t2$data_env, method = "spearman")
```
All the results are same. 
Please use `p_adjust_type = "Type"` if you think it is better.

Chi

**Comment by MPNguyen2022** [comment]
Hi, 

Thank you for your fast answer. I would like to clarify a bit if my understanding is correct. 

In the example above we have **9 alpha indices and 8 environmental variables**. So we have **72 pairs** of correlation.

So in microbiome package, the p-value was adjusted in a way:
**p.adjust(p, n=72)** ?
or p.adjust(p, n = 9) for each alpha index?

In microeco package: each p-value was adjusted with n = 9 (because we have 9 indices) for each environmental variable. Or was it adjusted with n = 8 (because we have 8 env. variables) for each alpha index?

Thank you again.

**Reply by ChiLiubio** [reply]
Yes. It depends on the choice of `p_adjust_type` parameter. The default is `Env`, i.e. adjust for each environmental variable in t1$res_cor seperately. If `Type`, adjustment is done according to the `Type` column in t1$res_cor, i.e. for all the data.

**Reply by MPNguyen2022** [reply]
Ok, thanks a lot. It would be great if you could specify these details in the manual. The current manual does not say much about this.
Best wishes.

**Reply by ChiLiubio** [reply]
Yes. I will update it! Many thanks!

**Comment by MPNguyen2022** [comment]
Hi,

I would like to ask what cal_cor does with the taxa abundance table before doing the correlation test. 

For example, my input data is untransformed OTU data.

At least I know that the command microbiome::associate does not do any transformation or filtering for the OTU table before the correlation test.

**Could you please explain what the microeco::cal_cor actually does to the OTU table (relative abundance transformation? merge by Genus level? filtering taxa?)?** 

Actually, I would like to do a correlation test for **relative abundance at Genus level to an environmental matrix**, but I don't know how to do it properly with microeco. 

Thank you very much.

**Comment by ChiLiubio** [comment, marked as answer]
Hi. correlation test for relative abundance at Genus level can be done like this easily.
```
library(microeco)
t1 <- trans_env$new(dataset = dataset, add_data = env_data_16S[, 4:11])
t1$cal_cor(use_data = "Genus", p_adjust_method = "fdr", cor_method = "spearman", p_adjust_type = "Type")
View(t1$res_cor)
```
The `cal_cor` function has no matter with the abundance transformation. It just invoke the Genus table in dataset$taxa_abund list. Actually, all the analysis referred to the abundance comes from the dataset$taxa_abund.
![图片](https://user-images.githubusercontent.com/20815519/224283373-1724d8e7-e901-447c-ba4f-551de749f5ed.png)
If you do not want to use relative abundance, you can recalculate the abundance with the parameter `rel`. It means the results in each table of dataset$taxa_abund is only the sum of otu_table at each taxonomic level.
```
dataset$cal_abund(rel = FALSE)
t1 <- trans_env$new(dataset = dataset, add_data = env_data_16S[, 4:11])
t1$cal_cor(use_data = "Genus", p_adjust_method = "fdr", cor_method = "spearman", p_adjust_type = "Type")
```
So the abundance (relative or absolute) calculation is in cal_abund function of microtable object. Other uses distribute in different classes. They are two differenct steps in order to be more flexible.
